lexicon/bpe.py

```diff
@@ -0,0 +1,141 @@
+__all__ = ["CreateBPELexiconJob"]
+
+import subprocess as sp
+import os
+import sys
+from typing import List, Optional, Set, Union
+import xml.etree.ElementTree as ET
+
+from sisyphus import Job, Task, tk
+
+from i6_core.lib.lexicon import Lexicon, Lemma
+import i6_core.util as util
+
+
+class CreateBPELexiconJob(Job):
+    """
+    In a Bliss lexicon replace the phonetic representation with a BPE decomposition of the words that can be used e.g, for lexicon constrained BPE search.
+
+    This job is still in experimental state, and only tested with Flashlight BPE decoding
+    """
+
+    def __init__(
+        self,
+        base_lexicon_path: tk.Path,
+        bpe_codes: tk.Path,
+        bpe_vocab: tk.Path,
+        subword_nmt_repo: tk.Path,
+        unk_label: str = "UNK",
+        vocab_blacklist: Optional[Union[List[str], Set[str]]] = None,
+        keep_special_lemmas: bool = True,
+    ):
+        """
+        :param base_lexicon_path: base lexicon (can be phoneme based) to take the lemmas from
+        :param bpe_codes: bpe codes from the ReturnnTrainBPEJob
+        :param bpe_vocab: vocab file to limit which bpe splits can be created
+        :param subword_nmt_repo: cloned repository
+        :param unk_label: unknown label, used in case a BPE token is created that is not in the vocab.
+        :param vocab_blacklist: which bpe_vocab entries not to load into the "phoneme/bpe-token" inventory
+            e.g. remove "<s>" and "</s>"
+        :param keep_special_lemmas: If special lemmas should be kept,
+            usually yes for RASR search and no for Flashlight search.
+            The phonemes of the special lemmas will also be kept, therefore
+            make sure there is no overlap with the BPE vocab.
+        """
+        self.base_lexicon_path = base_lexicon_path
+        self.bpe_codes = bpe_codes
+        self.bpe_vocab = bpe_vocab
+        self.subword_nmt_repo = subword_nmt_repo
+        self.unk_label = unk_label
+        if vocab_blacklist is None:
+            self.vocab_blacklist = set()
+        else:
+            # convert list to set for faster "in" check
+            self.vocab_blacklist = set(vocab_blacklist)
```

Comment:
Not very important but I don't think this is helpful. For short lists the "in" check is just as fast or even faster than for sets.

Reply:
But for short lists the speed is then not relevant, right? I am not sure if someone would ever put very large blacklists here, but I thought set is more consistent. I can change the comment though.